File: client/java/test/krpc/client/ConnectionTestCase.java
package krpc.client;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertThrows;

import java.io.IOException;
import java.nio.channels.ServerSocketChannel;
import krpc.schema.KRPC;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

/**
 * What a connection to a server does regardless of what carries it: making the handshake,
 * carrying a call, and reporting a server that turns the connection down. Only opening the
 * socket differs between the transports, so each of them supplies that and shares these.
 */
public abstract class ConnectionTestCase {

  private StandInServer server;

  /**
   * A channel listening on the transport under test, bound to somewhere nothing else is using.
   *
   * @return The listening channel.
   */
  protected abstract ServerSocketChannel listen() throws IOException;

  /**
   * Connect to the running server over the transport under test.
   *
   * @param name The name of the client.
   *
   * @return A connection to the server.
   */
  protected abstract Connection connect(String name) throws IOException;

  /**
   * The running server, for a test that needs to know where it is listening.
   *
   * @return The server.
   */
  protected StandInServer getServer() {
    return server;
  }

  @Before
  @SuppressWarnings("checkstyle:missingjavadocmethod")
  public void setup() throws IOException {
    server = new StandInServer(listen());
  }

  @After
  @SuppressWarnings("checkstyle:missingjavadocmethod")
  public void teardown() throws IOException {
    if (server != null) {
      server.close();
    }
  }

  @Test
  @SuppressWarnings("checkstyle:missingjavadocmethod")
  public void testConnectAndDisconnect() throws IOException {
    try (Connection connection = connect("JavaConnectionTest")) {
      assertNotNull(connection);
    }
  }

  @Test
  @SuppressWarnings("checkstyle:missingjavadocmethod")
  public void testCarriesOneCall() throws IOException, RPCException {
    // The stand-in answers every request, so this is the request reaching it and the response
    // coming back rather than anything the call means
    try (Connection connection = connect("JavaConnectionTest")) {
      assertNotNull(connection.invoke("TestService", "TestProcedure"));
    }
  }

  @Test
  @SuppressWarnings("checkstyle:missingjavadocmethod")
  public void testCarriesManyCalls() throws IOException, RPCException {
    // Calls are read out of a buffer the transport fills, so a run of them covers that buffer
    // being refilled and reused rather than only its first use
    try (Connection connection = connect("JavaConnectionTest")) {
      for (int i = 0; i < 100; i++) {
        assertNotNull(connection.invoke("TestService", "TestProcedure"));
      }
    }
  }

  @Test
  @SuppressWarnings("checkstyle:missingjavadocmethod")
  public void testConnectionRefusedByTheServer() throws IOException {
    server.refuseWith(KRPC.ConnectionResponse.Status.WRONG_TYPE,
        "Connection request was for the wrong server");
    ConnectionException exn =
        assertThrows(ConnectionException.class, () -> connect("JavaConnectionTestRefused"));
    assertEquals("Connection request was for the wrong server", exn.getMessage());
  }
}

File: client/csharp/test/ConnectionTestCase.cs
using System;
using System.Net.Sockets;
using KRPC.Schema.KRPC;
using NUnit.Framework;

namespace KRPC.Client.Test
{
    /// <summary>
    /// What a connection to a server does regardless of what carries it: making the
    /// handshake, carrying a call, and reporting a server that turns the connection down.
    /// Only opening the socket differs between the transports, so each of them supplies
    /// that and shares these.
    /// </summary>
    public abstract class ConnectionTestCase
    {
        StandInServer server;

        /// <summary>
        /// A socket listening on the transport under test, bound to somewhere nothing else
        /// is using.
        /// </summary>
        protected abstract Socket Listen ();

        /// <summary>
        /// Connect to the running server over the transport under test.
        /// </summary>
        protected abstract Connection Connect (string name);

        /// <summary>
        /// Where the running server is listening, for a test that opens its own socket.
        /// </summary>
        protected StandInServer Server {
            get { return server; }
        }

        [SetUp]
        public void SetUp ()
        {
            server = new StandInServer (Listen ());
        }

        [TearDown]
        public void TearDown ()
        {
            server.Dispose ();
        }

        [Test]
        public void ConnectAndDisconnect ()
        {
            using (var connection = Connect ("CSharpConnectionTest")) {
                Assert.IsNotNull (connection);
            }
        }

        [Test]
        public void CarriesACall ()
        {
            // The stand-in answers every request, so this measures the request reaching it and
            // the response coming back, and not what the call means
            using (var connection = Connect ("CSharpConnectionTest")) {
                Assert.IsNotNull (connection.Invoke ("TestService", "TestProcedure"));
            }
        }

        [Test]
        public void CarriesManyCalls ()
        {
            // A response is taken out of a buffer that reads fill a block at a time, so a run
            // of calls covers the buffer being refilled and reused, and not only its first use
            using (var connection = Connect ("CSharpConnectionTest")) {
                for (int i = 0; i < 100; i++)
                    Assert.IsNotNull (connection.Invoke ("TestService", "TestProcedure"));
            }
        }

        [Test]
        public void ConnectionRefusedByTheServer ()
        {
            server.Status = ConnectionResponse.Types.Status.WrongType;
            server.Message = "Connection request was for the wrong server";
            var exn = Assert.Throws<ConnectionException> (
                          () => Connect ("CSharpConnectionTestRefused"));
            Assert.AreEqual ("Connection request was for the wrong server", exn.Message);
        }
    }
}
